Sources/X509/TinyArray.swift

```diff
@@ -0,0 +1,230 @@
+    init() {
+        self = .arbitary([])
+    }
+    
+    @inlinable
+    mutating func append(_ newElement: Element) {
+        self.append(contentsOf: CollectionOfOne(newElement))
+    }
+    
+    @inlinable
+    mutating func append(contentsOf newElements: some Sequence<Element>){
+        switch self {
+        case .one(let firstElement):
+            var iterator = newElements.makeIterator()
+            guard let secondElement = iterator.next() else {
+                // newElements is empty, nothing to do
+                return
+            }
+            var elements: [Element] = []
+            elements.reserveCapacity(1 + newElements.underestimatedCount)
+            elements.append(firstElement)
+            elements.append(secondElement)
+            elements.appendRemainingElements(from: &iterator)
+            self = .arbitary(elements)
+            
+        case .arbitary(var elements):
+            if elements.isEmpty {
+                if let array = newElements as? Array<Element> {
+                    if array.count == 1 {
+                        self = .one(array[0])
+                    } else {
+                        self = .arbitary(array)
+                    }
+                } else {
+                    // if `self` is currently empty and `newElements` just contains a single
+                    // element, we skip allocating an array and set `self` to `.one(firstElement)`
+                    var iterator = newElements.makeIterator()
+                    guard let firstElement = iterator.next() else {
+                        // newElements is empty, nothing to do
+                        return
+                    }
+                    guard let secondElement = iterator.next() else {
+                        // newElements just contains a single element
+                        // and we hit the fast path
+                        self = .one(firstElement)
+                        return
+                    }
+                    elements.reserveCapacity(elements.count + newElements.underestimatedCount)
+                    elements.append(firstElement)
+                    elements.append(secondElement)
+                    elements.appendRemainingElements(from: &iterator)
+                    self = .arbitary(elements)
+                }
+            } else {
+                elements.append(contentsOf: newElements)
+                self = .arbitary(elements)
+            }
+            
+        }
+    }
+    
+    @discardableResult
+    @inlinable
+    mutating func remove(at index: Int) -> Element {
+        switch self {
+        case .one(let oldElement):
+            guard index == 0 else {
+                fatalError("index \(index) out of bounds")
+            }
+            self = .arbitary([])
+            return oldElement
+            
+        case .arbitary(var elements):
+            defer {
+                if elements.count == 1 {
+                    self = .one(elements[0])
```

Comment:
I'm not sure this is a worthwhile optimisation if we're aiming to reduce allocations: if `elements.count == 1` then we _must_ have had some heap storage before. If we drop batch to `.one` and later add another element then we'll be forced to allocate, if we keep the `.arbitrary` storage then we won't allocate when adding the next element. This would of course require you to implement `Hashable`/`Equatable`.

Reply:
I have wondered as well if we should do this or not. I have now implemented a custom Equatable/Hashable function and no longer throw the array away and instead store it.